pulser/simulation/simulation.py

```diff
@@ -15,116 +15,214 @@
+        if not (0 < value <= 1.0):
+            raise ValueError("`sampling_rate` must be positive and "
+                             "not larger than 1.0")
+        if int(self._tot_duration * value) < 4:
+            raise ValueError("`sampling_rate` is too small, less than 4 data "
+                             "points.")
+        self._sampling_rate: float = value
 
-        self._extract_samples()
-        self._build_basis_and_op_matrices()
-        self._construct_hamiltonian()
+    @property
+    def seq(self) -> Sequence:
+        """Property getter for sequence."""
+        return self._seq
+
+    @seq.setter
+    def seq(self, seq: Sequence) -> None:
+        if not isinstance(seq, Sequence):
+            raise TypeError("The provided sequence has to be a valid "
+                            "pulser.Sequence instance.")
+        if not seq._schedule:
+            raise ValueError("The provided sequence has no declared channels.")
+        if all(seq._schedule[x][-1].tf == 0 for x in seq._channels):
+            raise ValueError("No instructions given for the channels in the "
+                             "sequence.")
+        self._seq = seq
+
+    @property
+    def config(self) -> SimConfig:
+        """Property getter for config."""
+        return self._config
+
+    @config.setter
+    def config(self, cfg: SimConfig) -> None:
+        self._config = cfg
+        self._set_param_from_config()
+
+    def show_config(self) -> None:
+        """Shows current configuration."""
+        print(self._config)
+
+    def reset_config(self) -> None:
+        """Resets configuration to default."""
+        self.config = SimConfig()
+        print("Configuration has been set to default.")
 
-        if isinstance(evaluation_times, str):
-            if evaluation_times == 'Full':
-                self.eval_times = deepcopy(self._times)
-            elif evaluation_times == 'Minimal':
-                self.eval_times = np.array([self._times[0], self._times[-1]])
+    def _set_param_from_config(self) -> None:
+        """Sets all relevant Simulation parameters from its SimConfig.
+
+        Called every time a new configuration is loaded in order to update
+        Simulation parameters.
+        """
+        self._prepare_bad_atoms()
+        # update samples with potential noise settings
+        self._build_hamiltonian_from_seq()
+        if 'dephasing' in self.config.noise:
+            self._init_dephasing()
+
+    def _prepare_bad_atoms(self) -> None:
+        """Chooses atoms that will be badly prepared.
+
+        If no SPAM noise, simply sets all atoms as well-prepared.
+        """
+        # Tag True if atom is badly prepared
+        self._bad_atoms: dict[Union[str, int], bool] = {
+            qid: False for qid in self._qid_index}
+        if 'SPAM' in self.config.noise:
+            self._bad_atoms = \
+                {qid: (np.random.uniform() < self.config.spam_dict['eta'])
+                 for qid in self._qid_index}
```

Comment:
I have a suggestion here, for the sake of paralellization:
```suggestion
            dist = np.random.uniform(size=len(self._qid_index)) < self.config.spam_dict['eta']
            self._bad_atoms = dict(zip(self._qid_index, dist)) 
```

Reply:
I guess that works too, even though I don't see why this parallelizes better ?

Reply:
It's because you only `np.random.uniform` once. Numpy functions are built for parallelization. For example:
- `np.random.uniform(size=1000)`: 13.4 µs ± 578 ns
- `for _ in range(1000): np.random.uniform()` :  2.11 ms ± 70.1 µs

Reply:
Interesting, thanks for the heads-up !